call-home/src/bin/callhome/collector/report_models.rs

```diff
@@ -184,3 +199,217 @@ fn get_pools_size_vector(pools: Vec<openapi::models::Pool>) -> Vec<u64> {
+    let retry_policy = ExponentialBackoff::builder().build_with_max_retries(20);
+    let client_config = reqwest::Client::new();
+    ClientBuilder::new(client_config)
+        .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+        .build()
+}
+
+struct Record(Sample);
+
+impl Record {
+    fn action(&self) -> &str {
+        self.0.labels.get(ACTION).unwrap_or_default()
+    }
+
+    fn metric(&self) -> Option<Metrics> {
+        self.0.metric.clone().try_into().ok()
+    }
+
+    fn gauge_value(&self) -> u32 {
+        match self.0.value {
+            Value::Gauge(v) => v as u32,
+            _ => 0,
+        }
+    }
+}
+
+impl TryFrom<Record> for VolumeCreated {
+    type Error = ();
+    fn try_from(src: Record) -> Result<Self, ()> {
+        match src.metric() {
+            Some(Metrics::Volume) => {
+                if src.action() == "created" {
+                    Ok(VolumeCreated(src.gauge_value()))
+                } else {
+                    Err(())
+                }
+            }
+            _ => Err(()),
+        }
+    }
+}
+
+impl TryFrom<Record> for VolumeDeleted {
+    type Error = ();
+    fn try_from(src: Record) -> Result<Self, ()> {
+        match src.metric() {
+            Some(Metrics::Volume) => {
+                if src.action() == "deleted" {
+                    Ok(VolumeDeleted(src.gauge_value()))
+                } else {
+                    Err(())
+                }
+            }
+            _ => Err(()),
+        }
+    }
+}
+
+impl TryFrom<Record> for PoolCreated {
+    type Error = ();
+    fn try_from(src: Record) -> Result<Self, ()> {
+        match src.metric() {
+            Some(Metrics::Pool) => {
+                if src.action() == "created" {
+                    Ok(PoolCreated(src.gauge_value()))
+                } else {
+                    Err(())
+                }
+            }
+            _ => Err(()),
+        }
+    }
+}
+
+impl TryFrom<Record> for PoolDeleted {
+    type Error = ();
+    fn try_from(src: Record) -> Result<Self, ()> {
+        match src.metric() {
+            Some(Metrics::Pool) => {
+                if src.action() == "deleted" {
+                    Ok(PoolDeleted(src.gauge_value()))
+                } else {
+                    Err(())
+                }
+            }
+            _ => Err(()),
+        }
+    }
+}
+
+/// Metrics category.
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Metrics {
+    Pool,
+    Volume,
+    Unknown,
+}
+
+impl From<std::string::String> for Metrics {
+    fn from(s: std::string::String) -> Self {
+        match s.as_str() {
+            "pool" => Metrics::Pool,
+            "volume" => Metrics::Volume,
+            _ => Metrics::Unknown,
+        }
+    }
+}
+
+impl<T: TryFrom<Record>> From<&EventsRecord> for Option<T> {
+    fn from(src: &EventsRecord) -> Option<T> {
+        src.record
+            .iter()
+            .find_map(|r| T::try_from(Record(r.clone())).ok())
+    }
+}
```

Comment:
I think we can simplify this a lot by using a macro to generate the counters.
Also let's add an enum for Action and perhaps rename Metrics to Category (I'm on the fence on rename so either is fine)? 

```suggestion
struct Record<'a>(&'a Sample);
impl Record<'_> {
    fn action(&self) -> Action {
        Action::from(self.0.labels.get(ACTION).unwrap_or(""))
    }
    fn category(&self) -> Category {
        Category::from(self.0.metric.as_str())
    }
    fn gauge_value(&self) -> u32 {
        match self.0.value {
            Value::Gauge(v) => v as u32,
            _ => 0,
        }
    }
    fn try_gauge(&self, category: Category, action: Action) -> Result<GaugeCounter, ()> {
        if self.category() == category && self.action() == action {
            return Ok(GaugeCounter(self.gauge_value()));
        }
        Err(())
    }
}

/// Metrics category.
#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
pub enum Category {
    #[default]
    Unknown,
    Pool,
    Volume,
}
/// Action category.
#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
pub enum Action {
    #[default]
    Unknown,
    Created,
    Deleted,
}

impl From<&str> for Category {
    fn from(s: &str) -> Self {
        match s {
            "pool" => Category::Pool,
            "volume" => Category::Volume,
            _ => Category::Unknown,
        }
    }
}
impl From<&str> for Action {
    fn from(s: &str) -> Self {
        match s {
            "created" => Action::Created,
            "deleted" => Action::Deleted,
            _ => Action::Unknown,
        }
    }
}

macro_rules! make_gauge {
    ($name:ident, $category:expr, $action:expr) => {
        #[derive(Debug, Default, Clone)]
        pub struct $name(GaugeCounter);
        impl std::ops::Deref for $name {
            type Target = GaugeCounter;

            fn deref(&self) -> &Self::Target {
                &self.0
            }
        }
        impl TryFrom<Record<'_>> for $name {
            type Error = ();
            fn try_from(record: Record<'_>) -> Result<Self, ()> {
                record.try_gauge($category, $action).map($name)
            }
        }
    };
}

/// A gauge represents a single numerical value that can arbitrarily go up and down.
#[derive(Debug, Default, Clone)]
pub struct GaugeCounter(u32);
impl GaugeCounter {
    /// Get the inner value.
    pub fn value(&self) -> u32 {
        self.0
    }
}

make_gauge!(VolumeCreated, Category::Volume, Action::Created);
make_gauge!(VolumeDeleted, Category::Volume, Action::Deleted);
make_gauge!(PoolCreated, Category::Pool, Action::Created);
make_gauge!(PoolDeleted, Category::Pool, Action::Deleted);

impl<'a, T: TryFrom<Record<'a>>> From<&'a EventsRecord> for Option<T> {
    fn from(src: &'a EventsRecord) -> Option<T> {
        src.record.iter().find_map(|s| T::try_from(Record(s)).ok())
    }
}
```